### PR TITLE
Add support for 1.20.5/1.20.6 part 2

### LINKED ISSUE #2900: Paper 1.20.5/6 Failed
[08:49:56] [Server thread/ERROR]:   [ProtocolLib] INTERNAL ERROR: Cannot load ProtocolLib.
  If this problem hasn't already been reported, please open a ticket
  at https://github.com/dmulloy2/ProtocolLib/issues with the following data:
  Stack Trace:
  java.lang.RuntimeException: Failed to find NMS class: network.ProtocolInfo$a
  	at ProtocolLib(1).jar//com.comphenix.protocol.utility.MinecraftReflection.lambda$getMinecraftClass$2(MinecraftReflection.java:1383)
  	at java.base/java.util.Optional.orElseThrow(Unknown Source)
  	at ProtocolLib(1).jar//com.comphenix.protocol.utility.MinecraftReflection.getMinecraftClass(MinecraftReflection.java:1383)
  	at ProtocolLib(1).jar//com.comphenix.protocol.injector.packet.PacketRegistry.createRegisterV1_20_5(PacketRegistry.java:344)
  	at ProtocolLib(1).jar//com.comphenix.protocol.injector.packet.PacketRegistry.initialize(PacketRegistry.java:498)
  	at ProtocolLib(1).jar//com.comphenix.protocol.injector.packet.PacketRegistry.getClientPacketTypes(PacketRegistry.java:538)
  	at ProtocolLib(1).jar//com.comphenix.protocol.injector.PacketFilterManager.<init>(PacketFilterManager.java:120)
  	at ProtocolLib(1).jar//com.comphenix.protocol.injector.PacketFilterBuilder.build(PacketFilterBuilder.java:121)
  	at ProtocolLib(1).jar//com.comphenix.protocol.ProtocolLib.onLoad(ProtocolLib.java:183)
  	at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:59)
  	at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:18)
  	at io.papermc.paper.plugin.storage.SimpleProviderStorage.enter(SimpleProviderStorage.java:39)
  	at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enter(LaunchEntryPointHandler.java:36)
  	at org.bukkit.craftbukkit.CraftServer.loadPlugins(CraftServer.java:508)
  	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:287)
  	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1140)
  	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:326)
  	at java.base/java.lang.Thread.run(Unknown Source)
  Dump:
  Parameters: 
    [NULL]
  Sender:
    com.comphenix.protocol.ProtocolLib@42b30cfe[
      statistics=<null>
      packetTask=<null>
      tickCounter=0
      configExpectedMod=-1
      updater=com.comphenix.protocol.updater.SpigotUpdater@56e53e50
      redirectHandler=<null>
      scheduler=com.comphenix.protocol.scheduler.DefaultScheduler@ceb0c8e
      commandProtocol=<null>
      commandPacket=<null>
      commandFilter=<null>
      packetLogging=<null>
      skipDisable=false
      isEnabled=false
      loader=io.papermc.paper.plugin.manager.DummyBukkitPluginLoader@2a764c61
      server=CraftServer{serverName=Paper,serverVersion=git-Paper-6,minecraftVersion=1.20.5}
      file=plugins/.paper-remapped/ProtocolLib(1).jar
      description=org.bukkit.plugin.PluginDescriptionFile@10c8f01e
      pluginMeta=org.bukkit.plugin.PluginDescriptionFile@10c8f01e
      dataFolder=plugins/ProtocolLib
      classLoader=PluginClassLoader{plugin=ProtocolLib v5.2.1-SNAPSHOT-686, pluginEnabled=false, url=plugins/.paper-remapped/ProtocolLib(1).jar}
      naggable=true
      newConfig=YamlConfiguration[path='', root='YamlConfiguration']
      configFile=plugins/ProtocolLib/config.yml
      logger=com.destroystokyo.paper.utils.PaperPluginLogger@53a5cd2f
      lifecycleEventManager=io.papermc.paper.plugin.lifecycle.event.PaperLifecycleEventManager@487a2487
      allowsLifecycleRegistration=true
    ]
  Version:
    ProtocolLib v5.2.1-SNAPSHOT-686
  Java Version:
    21.0.3
  Server:
    git-Paper-6 (MC: 1.20.5)

### LINKED ISSUE #2903: 1.20.5
Its not working

### LINKED ISSUE #2907: 1.20.6
Paper 1.20.6 dev build #35

Error from protocollib (also the reason PL-HIDE-PRO doesnt work now)

https://pastebin.com/yk0ZvPmQ

### PR DESCRIPTION
**Pull request by Ingrim4**
## Description
See #2894

## Related Issue
closes #2907
closes #2900
closes #2903

## How Has This Been Tested?
Untested as of yet.

## Types of Changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by ClumsyIsNotReal** [comment]
Compiled version with the changes and still getting errors using Latest Paper 1.20.6. Paper uses Mojang mapping which could be the cause between the differences of it working on spigot and not paper

`[19:59:57` INFO]: [ProtocolLib] Loading server plugin ProtocolLib v5.2.1-SNAPSHOT
[19:59:57 WARN]: [ProtocolLib] Version (MC: 1.20.6) has not yet been tested! Proceed with caution.
[19:59:58 ERROR]: [ProtocolLib] Assuming package version: v1_20_R1
[19:59:58 ERROR]:   [ProtocolLib] INTERNAL ERROR: Cannot load ProtocolLib.
  If this problem hasn't already been reported, please open a ticket
  at https://github.com/dmulloy2/ProtocolLib/issues with the following data:
  Stack Trace:
  java.lang.RuntimeException: Failed to find NMS class: network.ProtocolInfo$a
        at ProtocolLib.jar//com.comphenix.protocol.utility.MinecraftReflection.lambda$getMinecraftClass$2(MinecraftReflection.java:1383)
        at java.base/java.util.Optional.orElseThrow(Optional.java:403)
        at ProtocolLib.jar//com.comphenix.protocol.utility.MinecraftReflection.getMinecraftClass(MinecraftReflection.java:1383)
        at ProtocolLib.jar//com.comphenix.protocol.injector.packet.PacketRegistry.createRegisterV1_20_5(PacketRegistry.java:344)
        at ProtocolLib.jar//com.comphenix.protocol.injector.packet.PacketRegistry.initialize(PacketRegistry.java:498)
        at ProtocolLib.jar//com.comphenix.protocol.injector.packet.PacketRegistry.getClientPacketTypes(PacketRegistry.java:538)
        at ProtocolLib.jar//com.comphenix.protocol.injector.PacketFilterManager.<init>(PacketFilterManager.java:120)
        at ProtocolLib.jar//com.comphenix.protocol.injector.PacketFilterBuilder.build(PacketFilterBuilder.java:121)
        at ProtocolLib.jar//com.comphenix.protocol.ProtocolLib.onLoad(ProtocolLib.java:183)
        at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:59)
        at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:18)
        at io.papermc.paper.plugin.storage.SimpleProviderStorage.enter(SimpleProviderStorage.java:39)
        at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enter(LaunchEntryPointHandler.java:36)
        at org.bukkit.craftbukkit.CraftServer.loadPlugins(CraftServer.java:508)
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:287)
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1140)
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:326)
        at java.base/java.lang.Thread.run(Thread.java:1583)
  Dump:
  Parameters: 
    [NULL]
  Sender:
    com.comphenix.protocol.ProtocolLib@7723e60a[
      statistics=<null>
      packetTask=<null>
      tickCounter=0
      configExpectedMod=-1
      updater=com.comphenix.protocol.updater.SpigotUpdater@1c138bfd
      redirectHandler=<null>
      scheduler=com.comphenix.protocol.scheduler.DefaultScheduler@74683dc3
      commandProtocol=<null>
      commandPacket=<null>
      commandFilter=<null>
      packetLogging=<null>
      skipDisable=false
      isEnabled=false
      loader=io.papermc.paper.plugin.manager.DummyBukkitPluginLoader@70864bf9
      server=CraftServer{serverName=Paper,serverVersion=git-Paper-48,minecraftVersion=1.20.6}
      file=plugins/.paper-remapped/ProtocolLib.jar
      description=org.bukkit.plugin.PluginDescriptionFile@11b502f7
      pluginMeta=org.bukkit.plugin.PluginDescriptionFile@11b502f7
      dataFolder=plugins/ProtocolLib
      classLoader=PluginClassLoader{plugin=ProtocolLib v5.2.1-SNAPSHOT, pluginEnabled=false, url=plugins/.paper-remapped/ProtocolLib.jar}
      naggable=true
      newConfig=YamlConfiguration[path='', root='YamlConfiguration']
      configFile=plugins/ProtocolLib/config.yml
      logger=com.destroystokyo.paper.utils.PaperPluginLogger@1c7ec9f2
      lifecycleEventManager=io.papermc.paper.plugin.lifecycle.event.PaperLifecycleEventManager@74bd6ce9
      allowsLifecycleRegistration=true
    ]
  Version:
    ProtocolLib v5.2.1-SNAPSHOT
  Java Version:
    21.0.3
  Server:
    git-Paper-48 (MC: 1.20.6)
[19:59:58 INFO]: Server permissions file permissions.yml is empty, ignoring it
[19:59:58 INFO]: [ProtocolLib] Enabling ProtocolLib v5.2.1-SNAPSHOT
[19:59:58 ERROR]: ******************************************************
[19:59:58 ERROR]: *** ProtocolLib does not support plugin reloaders! ***
[19:59:58 ERROR]: *** Please use the built-in reload command!        ***
[19:59:58 ERROR]: ******************************************************
[19:59:58 INFO]: [ProtocolLib] Disabling ProtocolLib v5.2.1-SNAPSHOT``

**Comment by Ingrim4** [comment]
> Compiled version with the changes and still getting errors using Latest Paper 1.20.6. Paper uses Mojang mapping which could be the cause between the differences of it working on spigot and not paper

Works fine for me, did you compile the code from this PR or some older code?

**Comment by ClumsyIsNotReal** [comment]
> > Compiled version with the changes and still getting errors using Latest Paper 1.20.6. Paper uses Mojang mapping which could be the cause between the differences of it working on spigot and not paper
> 
> Works fine for me, did you compile the code from this PR or some older code?

Could you add me on discord: clumsyisnotreal

make it easier to see what is up

**Comment by nicholas-gamble** [comment]
No luck on my end either.
```
[11:26:19 INFO]: [ProtocolLib] Loading server plugin ProtocolLib v5.2.1-SNAPSHOT
[11:26:19 ERROR]: [ProtocolLib] Error initializing plugin 'ProtocolLib.jar' in folder 'plugins/.paper-remapped' (Is it up to date?)
java.lang.NoClassDefFoundError: net/bytebuddy/dynamic/scaffold/subclass/ConstructorStrategy
        at ProtocolLib.jar/com.comphenix.protocol.ProtocolLib.onLoad(ProtocolLib.java:129) ~[ProtocolLib.jar:?]
        at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:59) ~[paper-1.20.6.jar:git-Paper-49]
        at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:18) ~[paper-1.20.6.jar:git-Paper-49]
        at io.papermc.paper.plugin.storage.SimpleProviderStorage.enter(SimpleProviderStorage.java:39) ~[paper-1.20.6.jar:git-Paper-49]
        at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enter(LaunchEntryPointHandler.java:36) ~[paper-1.20.6.jar:git-Paper-49]
        at org.bukkit.craftbukkit.CraftServer.loadPlugins(CraftServer.java:508) ~[paper-1.20.6.jar:git-Paper-49]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:287) ~[paper-1.20.6.jar:git-Paper-49]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1140) ~[paper-1.20.6.jar:git-Paper-49]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:326) ~[paper-1.20.6.jar:git-Paper-49]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.ClassNotFoundException: net.bytebuddy.dynamic.scaffold.subclass.ConstructorStrategy
        at org.bukkit.plugin.java.PluginClassLoader.loadClass0(PluginClassLoader.java:197) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.java.PluginClassLoader.loadClass(PluginClassLoader.java:164) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526) ~[?:?]
        ... 10 more
```

```
git log
commit d26cf5febc7e7e19b2c8b60985c6517f122d670d (HEAD, origin/feat-1.20.5-support-part2)
Author: Maurice Eisenblätter <code@ingrim4.me>
Date:   Sun May 5 22:18:17 2024 +0200

    chore: add some missing packet types
```

File: src/main/java/com/comphenix/protocol/utility/MinecraftMethods.java [hunk 1]
```diff
@@ -231,6 +253,11 @@ private static void initializePacket() {
                 }
             }
 
+            // write and read methods are no longer part of the packet interface since 1.20.5
+            if (MinecraftVersion.v1_20_5.atOrAbove()) {
```
Comment (dmulloy2) [inline src/main/java/com/comphenix/protocol/utility/MinecraftMethods.java:257]: do we want to throw an exception here?

File: src/test/java/com/comphenix/protocol/PacketTypeTest.java [hunk 2]
```diff
@@ -52,60 +51,28 @@ public static void beforeClass() {
...
 
-                    try {
-                        PacketType.fromClass(entry1.getValue());
-                    } catch (Exception ex) {
-                        newTypes.add(entry1.getValue());
-                    }
-                }
-            }
-        }
+        PacketType.onDynamicCreate = (type, className) -> System.out.printf("%s, %s = new PacketType(PROTOCOL, SENDER, %s, \"%s\")\n", type.getProtocol(), type.getSender(), formatHex(type.getCurrentId()), className);
```
Comment (dmulloy2) [inline src/test/java/com/comphenix/protocol/PacketTypeTest.java:62]: love it

**Review by dmulloy2 — APPROVED** [review]

**Comment by dmulloy2** [comment]
looks great! any particular reason it's still a draft?

**Comment by Ingrim4** [comment]
Still would like to change some stuff and add missing packet types.

File: src/main/java/com/comphenix/protocol/utility/MinecraftMethods.java (hunk 1 above)
Comment (Ingrim4) [reply src/main/java/com/comphenix/protocol/utility/MinecraftMethods.java:257]: Yeah I guess we could throw an exception when the read/write method getters get called on 1.20.5 upwards

File: src/test/java/com/comphenix/protocol/PacketTypeTest.java (hunk 2 above)
Comment (Ingrim4) [reply src/test/java/com/comphenix/protocol/PacketTypeTest.java:62]: Seemed like the easiest way to me I would also like to parse the packet names and only print the short name if possible:
e.g. `ServerboundClientInformationPacket` -> `ClientInformation`

**Comment by dmulloy2** [comment]
> Still would like to change some stuff and add missing packet types.

gotcha. well let me know when it's ready. even if it requires a follow up PR, seems like this would help with some of the errors people are seeing

**Comment by Ingrim4** [comment]
> gotcha. well let me know when it's ready. even if it requires a follow up PR, seems like this would help with some of the errors people are seeing

@dmulloy2 I think I'm done for now with everything 1.20.5/1.20.6 and packet related. Some wrappers probably won't work I also didn't test everything just some basic async packet listener.

Completely off-topic but related to may work on this PR: After delving into ProtocolLib's internal code I feel like certain aspects, like packet types (packet IDs), require a revamp to align better with the current state of the game. However, such changes might break compatibility with older game versions pre 1.7. That's why, it's uncertain whether such modifications are desirable for this project. My proposal is to introduce ProtocolLib 6.0.0, designed for Minecraft 1.16+ and Java 17+. This approach would enable us to overhaul significant components of the system while retaining support for the majority of versions still in use today. If that approach is to extreme I would still like some feedback on this since I would also be interested in a less extreme approach that tries to keep comparability with 1.8, 1.12 and 1.16+. Might also be a good idea to start a github discussion on this and get some feedback from other people.

**Comment by dmulloy2** [comment]
agreed. worth looking into mojang mapping as well

**Comment by GroovyNoah** [comment]
can u ad me on discord groovynoah.eth I have an issue with protocollib id like to discuss

**Pull request merged**